openssl/src/x509/sbgp.rs

```diff
@@ -0,0 +1,252 @@
+    }
+
+    /// Determines whether the IP address extension is contained
+    /// in the parents IP address extension.
+    #[corresponds(X509v3_addr_subset)]
+    pub fn subset_of(&self, parent: &Stack<IPAddressFamily>) -> bool {
+        unsafe { ffi::X509v3_addr_subset(self.as_ptr(), parent.as_ptr()) == 1 }
+    }
+
+    /// Returns whether the IP address extension is in canonical
+    /// form.
+    #[corresponds(X509v3_addr_is_canonical)]
+    pub fn is_canonical(&self) -> bool {
+        unsafe { ffi::X509v3_addr_is_canonical(self.as_ptr()) == 1 }
+    }
+}
+
+impl IPAddressFamily {
+    /// Returns the IP version of this record.
+    #[corresponds(X509v3_addr_get_afi)]
+    pub fn fam(&self) -> Option<IpVersion> {
+        unsafe {
+            match X509v3_addr_get_afi(self.as_ptr()) as libc::c_int {
+                IANA_AFI_IPV4 => Some(IpVersion::V4),
+                IANA_AFI_IPV6 => Some(IpVersion::V6),
+                _ => None,
+            }
+        }
+    }
+
+    /// Returns the list of IP ranges contained in this IP family.
+    #[corresponds(X509v3_addr_get_range)]
+    pub fn range(&self) -> Option<Vec<(IpAddr, IpAddr)>> {
+        unsafe {
+            let fam = self.fam()?;
+            let choice = (*self.as_ptr()).ipAddressChoice;
+            if (*choice).type_ != IPAddressChoice_addressesOrRanges {
+                return None;
+            }
+            let stack =
+                StackRef::<IPAddressOrRange>::from_const_ptr_opt((*choice).u.addressesOrRanges)?;
+            match fam {
+                IpVersion::V4 => Self::addr_get_range::<Ipv4Addr>(stack, 4, IANA_AFI_IPV4 as u32),
+                IpVersion::V6 => Self::addr_get_range::<Ipv6Addr>(stack, 16, IANA_AFI_IPV6 as u32),
+            }
+        }
+    }
+
+    unsafe fn addr_get_range<T: Into<IpAddr>>(
+        stack: &StackRef<IPAddressOrRange>,
+        len: i32,
+        afi: u32,
+    ) -> Option<Vec<(IpAddr, IpAddr)>> {
+        let mut r = Vec::new();
+        for val in stack {
+            let mut min = MaybeUninit::<T>::uninit();
+            let mut max = MaybeUninit::<T>::uninit();
+            let size = ffi::X509v3_addr_get_range(
+                val.as_ptr(),
+                afi,
+                min.as_mut_ptr() as *mut _,
+                max.as_mut_ptr() as *mut _,
+                len,
+            );
+            if size != len {
+                return None;
+            }
+            r.push((min.assume_init().into(), max.assume_init().into()))
+        }
+        Some(r)
+    }
+}
+
+impl X509Ref {
+    /// Extracts the SBGP AS number extension from the `X509` certificate.
+    #[corresponds(X509_get_ext_d2i)]
+    pub fn sbgp_asn(&self) -> Option<ASIdentifiers> {
+        unsafe {
+            let asn = ffi::X509_get_ext_d2i(
+                self.as_ptr(),
+                ffi::NID_sbgp_autonomousSysNum,
+                std::ptr::null_mut(),
+                std::ptr::null_mut(),
+            );
```

Comment:
`X509_get_ext_d2i` is a tricky API to use correctly. As far as I can see, no caller in rust-openssl gets it right.

This can return `NULL` for several reasons: the benign reason is that there is no extension corresponding to the nid in the cert. However, there could also be a library-internal error (malformed extension, memory allocation error, ...), or other errors, like multiple extensions of the same type, which this idiom silently ignores.

The correct way is to pass in a pointer `crit` to a `c_int` as third argument. If `NULL` is returned and `crit` is set to anything but `-1`, an error should be raised.

Not sure if this should be fixed in this PR or if rust-openssl wants to land this as-is and fix the issue in a sweep.

Reply:
After considering this further, if there would be a minimally invasive solution available, I'd opt to address it within this PR. However, it seems that we require some new Error construct due to the fact that X509_get_ext_d2i doesn't set the internal OpenSSL error, making the ErrorStack inapplicable.

I'm unsure if this issue extends to other OpenSSL functions as well, where a new Error construct would also help to handle return values properly. After all, I propose addressing this in a separate PR to avoid overloading the current one.

Reply:
In the meantime of the last few months we also got some additional X509 functionality.
Yet, I suggest following the same procedure: let's complete the current task first, and then I'll create an additional PR specifically for the new functions.

Reply:
@botovq

Reply:
@HolyShitMan I'm inclined to agree, but it isn't my call. I opened #2226, which, once addressed, should also fix the instance in this PR.

Reply:
@botovq: #2226 and it's solution proposed in #2240 have been open for over a month without any significant progress. Could we find a way to expedite this? I am keen to get this MR resolved.